git.go

```diff
@@ -207,14 +218,29 @@ func SetupRepoAuth(options *Options) transport.AuthMethod {
 	// Assume SSH auth for all other formats.
 	options.Logger(codersdk.LogLevelInfo, "#1: 🔑 Using SSH authentication!")
 
-	var signer ssh.Signer
+	var signer gossh.Signer
 	if options.GitSSHPrivateKeyPath != "" {
 		s, err := ReadPrivateKey(options.GitSSHPrivateKeyPath)
 		if err != nil {
 			options.Logger(codersdk.LogLevelError, "#1: ❌ Failed to read private key from %s: %s", options.GitSSHPrivateKeyPath, err.Error())
 		} else {
-			options.Logger(codersdk.LogLevelInfo, "#1: 🔑 Using %s key!", s.PublicKey().Type())
 			signer = s
+			options.Logger(codersdk.LogLevelInfo, "#1: 🔑 Using %s key %s!", s.PublicKey().Type(), keyFingerprint(signer)[:8])
+		}
+	}
+
+	// If we have no signer but we have a Coder URL and agent token, try to fetch
+	// an SSH key from Coder!
+	if signer == nil && options.CoderAgentURL != "" && options.CoderAgentToken != "" {
+		options.Logger(codersdk.LogLevelInfo, "#1: 🔑 Fetching key from %s!", options.CoderAgentURL)
+		fetchCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		s, err := FetchCoderSSHKeyRetry(fetchCtx, options.Logger, options.CoderAgentURL, options.CoderAgentToken)
+		if err == nil {
+			signer = s
+			options.Logger(codersdk.LogLevelInfo, "#1: 🔑 Fetched %s key %s !", signer.PublicKey().Type(), keyFingerprint(signer)[:8])
+		} else {
+			options.Logger(codersdk.LogLevelInfo, "#1: ❌ Failed to fetch SSH key from %s: %w", options.CoderAgentURL, err)
 		}
```

Comment:
Deeply integrating the hot-path code with Coder feels incorrect to me.

I don't believe the `envbuilder` package should have any knowledge of the Coder agent or Coder URL. Prior to the CLI change, it didn't. The command-line simply logged to Coder as a sink if those vars were provided, and I think implementing this similarly abstractly would be an improvement.

We want people outside of the Coder ecosystem to use envbuilder. By integrating the codebase so heavily into Coder, we alienate that possibility.

Reply:
My preference would be to allow passing SSH options to envbuilder, and also not provide Coder options directly to envbuilder.

Reply:
How about a pre-init hook command?

Reply:
Adopting "lifecycle hooks" is a reasonable concept, we may argue where they should be scripts or just shell commands. Maybe we can implement them _similarly_ to [Git Hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks)?

Reply:
Any thoughts as to what kind of hook, except pre-clone, we might need in envbuilder? Running scripts is generally already addressable by `devcontainer.json`.

Reply:
If we added a new data source to the coder provider, I believe we address this similarly to what @kylecarbs suggested:

```
container {
  env = {
    GIT_SSH_KEY: basy64(data.coder_workspace.owner_ssh_key)
  }
}
```

Reply:
That sounds like a wonderful way to do it!

Reply:
This is probably out of scope but I want to share my thoughts,

As the ssh key is per user and not per workspace we should probably add/refactor a new data source `coder_user`

```
container {
  env = {
    GIT_SSH_KEY: basy64(data.coder_owner.private_ssh_key)
  }
}
```

Others properties we can add or move from `coder_workspace` can be
1. `coder_workspace.owner` -> `coder_user.name`
2. `coder_workspace.owner_name` -> `coder_user.full_name`
3. `coder_workspace.owner_emial` -> `coder_user.email`
4. Add `coder_user.private_ssh_key`
5. Add `coder_user.public_ssh_key`
6. Add `coder_user.groups` (A list of all groups if applicable)

cc: @bpmct

Reply:
Yeah, I think you're right that another data source for the coder_user is good, but shouldn't block this effort.

Reply:
Mentioning that in https://github.com/coder/terraform-provider-coder/issues/217

Reply:
> If we added a new data source to the coder provider, I believe we can address this similarly to what @kylecarbs suggested:
> 
> ```terraform
> container {
>   env = {
>     GIT_SSH_KEY: basy64(data.coder_workspace.owner_ssh_key)
>   }
> }
> ```

👍 That's certainly a valid path forward.

Based on what was implemented in #170, there would be a few extra steps:
- Create a secret containing the user's private key
- Mount that secret into the container under a well-known path
- Reference that path in the container's environment

Will open a separate PR for this.

Reply:
Already moved. See
https://github.com/coder/terraform-provider-coder/issues/219